Update ADK doc according to issue #1350 - 6 (#1361)
google/adk-docs: docs/sessions/memory.md

Reviewer: The main deliverable for issue #1350 item 6 (the new `add_memory` method and `memories.create` vs `memories.generate consolidation behavior`) is missing from this PR.

Please bring in the proposed edits from PR #1358, specifically the `### Direct Memory Ingestion with add_memory` subsection into this PR, placing it here after the "Retrieving Memories" bullet. Then you can edit the content further as needed.

> # Memory: long-term knowledge with `MemoryService`

Reviewer: ```suggestion
# Memory: Long-term knowledge with `MemoryService`
```

</div>

While a `Session` tracks the history (`events`) and temporary data (`state`) of a single conversation, an agent may need to recall information from past interactions. This is where the concept of **Long-Term Knowledge** and the **`MemoryService`** come into play. Think of it this way:

* **`Session` / `State`:** It's your short-term memory during one specific chat.
* **Long-Term Knowledge (`MemoryService`)**: It's a searchable archive or knowledge library the agent can consult, potentially containing information from many past chats or other sources.

## The `MemoryService` role

The `BaseMemoryService` (or `Service` in Go) defines the interface for managing this searchable, long-term knowledge store. It supports four operations:

1. **Ingesting a session (`add_session_to_memory`):** Take the contents of a completed `Session` and add relevant information to the long-term knowledge store.
2. **Ingesting events incrementally (`add_events_to_memory`):** Append a delta of events, for example the latest turn, without re-ingesting the full session. Useful when you want to write to memory partway through a long-running session.

> 3. **Writing memory items directly (`add_memory`):** Insert pre-built `MemoryEntry` items, for services that support direct writes alongside event-based extraction.

Reviewer: #1358 proposed some different wording for these same bullet points. Please compare its version to this list and pull in any better wording, if you want to. The good news is that what is currently on `main` already covers these operations, and you can ignore #1358's list structure, so this is just a check to see if you want any of the wording from these bullets from #1358.

1. **Ingesting a session (`add_session_to_memory`):** Take the contents of a completed `Session` and add relevant information to the long-term knowledge store.
2. **Ingesting events incrementally (`add_events_to_memory`):** Append a delta of events, for example the latest turn, without re-ingesting the full session. Useful when you want to write to memory partway through a long-running session.
3. **Writing memory items directly (`add_memory`):** Insert pre-built `MemoryEntry` items, for services that support direct writes alongside event-based extraction.
4. **Searching (`search_memory`):** Allow an agent (typically via a `Tool`) to query the knowledge store and retrieve relevant snippets based on a search query.
4. **Searching (`search_memory`):** Allow an agent, typically via a `Tool`, to query the knowledge store and retrieve relevant snippets based on a search query.

Operations 2 and 3 are optional — the base class implementations of `add_events_to_memory` and `add_memory` raise `NotImplementedError`, so check your concrete service before relying on them.

> ## Choosing the Right Memory Service

Reviewer: Removing the `## Choosing the Right Memory Service` heading leaves the comparison table with no section heading and no TOC entry. It's a navigation downgrade, so I suggest keeping the heading here.

> | **Persistence** | None, data is lost on restart | Yes, managed by agent platform | Yes, stored in Knowledge Engine |

Reviewer: "Agent Platform" is the new name for "Vertex AI", and it should be Title Cased since it's a product name.

Reviewer 2: Noted, thank you!

> ## Advanced Concepts

Reviewer: ```suggestion
## Advanced concepts
```

### How memory works in practice

The memory workflow internally involves these steps:

> 1. **Session Interaction:** A user interacts with an agent via a `Session`, managed by a `SessionService`. During this interaction, events are recorded and session state is updated.

Reviewer: "Session state is updated" overstates the original conditional "state might be updated" (state isn't always updated). Suggest "session state may be updated."

```suggestion
1. **Session Interaction:** A user interacts with an agent via a `Session`, managed by a `SessionService`. During this interaction, events are recorded and session state may be updated.
```

2. **Ingestion into Memory:** When a session concludes or captures significant information, your application calls `memory_service.add_session_to_memory(session)`. This action extracts key data and persists it to your long-term knowledge store, such as the Agent Runtime Memory Bank.
3. **Later Query:** In a different, or in the same session, you might ask a question requiring past context, for example, "What did we discuss about project X last week?".
4. **Agent Uses Memory Tool:** An agent equipped with a memory-retrieval tool, such as the built-in `load_memory` tool, recognizes the need for past context. It calls the tool, providing a search query (e.g., "discussion project X last week").
5. **Search Execution:** The tool internally calls `memory_service.search_memory(app_name=..., user_id=..., query=...)`.

> 6. **Results Returned:** The `MemoryService` searches its store, using keyword matching or semantic search, and returns matching snippets as a `SearchMemoryResponse` containing a list of `MemoryEntry` objects, each holding `content`, and all optional: `author`,`timestamp`, and `custom_metadata`.

Reviewer: ```suggestion
6. **Results Returned:** The `MemoryService` searches its store, using keyword matching or semantic search, and returns matching snippets as a `SearchMemoryResponse` containing a list of `MemoryEntry` objects, each holding `content`, and all optional: `author`, `timestamp`, and `custom_metadata`.
```

7. **Agent Uses Results:** The tool returns these results to the agent, usually as part of the context or function response. The agent can then use this retrieved information to formulate its final answer to the user.

### Can an agent have access to more than one memory service?

*   **Through Standard Configuration: No.** The framework (`adk web`, `adk api_server`) is designed to be configured with one memory service at a time via the `--memory_service_uri` flag. That single service is wired into the runner and exposed through `tool_context.search_memory()` and `callback_context.search_memory()`.

> *   **Within Your Agent's Code: Yes, absolutely.** Manually import and instantiate additional memory services to meet your specific requirements. This approach enables your agent to access multiple memory sources within a single turn.

Reviewer: This rewrite drops the specific, useful guidance from the original ("the cleanest place to consult it is from a custom tool, which already has a ToolContext for the framework-configured service") and replaces it with vaguer text. Suggest keeping most of the original guidance since it's more actionable and shorter. And we should not use the form "Yes, absolutely" in docs.

```suggestion
*   **Within Your Agent's Code: Yes.** You can instantiate a second `BaseMemoryService` and consult it from a custom tool, which already has a `ToolContext` for the framework-configured service.
```
